packages/hoprd/crates/hoprd-inbox/src/ring.rs

```diff
@@ -143,16 +150,26 @@ where
 
         self.buffers
             .get_mut(&specific_tag)
-            .and_then(|buf| buf.peek().map(|w| (w.payload.clone(), w.ts)))
+            .and_then(|buf| 
+                buf
+                .peek()
+                .map(|w| (w.payload.clone(), w.ts)))
     }
 
-    async fn peek_all(&mut self, tag: Option<T>) -> Vec<(M, Duration)> {
+    async fn peek_all(&mut self, tag: Option<T>, timestamp: Option<Duration>) -> Vec<(M, Duration)> {
+        let timestamp = timestamp.unwrap_or(Duration::from_millis(0));
+
         match tag {
             Some(specific_tag) => {
                 // Peek only all messages of a specific tag
                 self.buffers
                     .get_mut(&specific_tag)
-                    .map(|buf| buf.iter().map(|w| (w.payload.clone(), w.ts)).collect::<Vec<_>>())
+                    .map(|buf| {
+                        buf.iter()
+                            .filter(|w| w.ts >= timestamp)
+                            .map(|w| (w.payload.clone(), w.ts))
```

Comment:
```suggestion
                            .filter_map(|w| if w.ts >= timestamp { Some((w.payload.clone(), w.ts)) } else { None } )
```

Reply:
Or just `(w.ts >= timestamp).then_some((w.payload.clone(), w.ts))` to avoid the if-else.

Reply:
In this case it does not matter, but it'd be preferred to limit the use of extensions with eagerly evaluated functionality.